### PR TITLE
Speed up Constraint Solver's overload resolution

### PR DESCRIPTION
**Pull request by gmpl**
This speeds up to 20x overload resolution and solves #343 at least the repro there mentioned.

This is what I get running the repro many times with the current version of the compiler:

    val compileTime10 : float = 162.0047
    val compileTime11 : float = 3463.0556
    val ratio11to10 : float = 21.37626624

    val compileTime10 : float = 156.0002
    val compileTime11 : float = 3039.0392
    val ratio11to10 : float = 19.48099554

    val compileTime10 : float = 171.0647
    val compileTime11 : float = 2965.9741
    val ratio11to10 : float = 17.33831761

And here's what I get with this optimized version:

    val compileTime10 : float = 165.0014
    val compileTime11 : float = 180.0049
    val ratio11to10 : float = 1.090929532

    val compileTime10 : float = 158.0022
    val compileTime11 : float = 182.0045
    val ratio11to10 : float = 1.151911176

    val compileTime10 : float = 209.0025
    val compileTime11 : float = 205.0066
    val ratio11to10 : float = 0.9808810899

It also seems logic to me that when exploring each possible solution, the constraint should be assumed as solved. I can't imagine a reason why not to assume that.

In case there is interest I have more complex repros with very interesting results.

### CODE REVIEW

**Comment by dsyme** [comment]
Super, super interesting, especially since the adjustments are small 

Seeing these failures - not sure why - some changes in errors/warnings may be expected but there are other failures here too
```
1) Failed : FSharp-Tests-Core+Members+Basics.Basics(GENERATED_SIGNATURE)
Error running command 'D:\j\workspace\release_ci_pa---866fd2c3\tests\..\release\net40\bin\fsc.exe' with args ' -r:System.Core.dll --nowarn:20 --define:COMPILED --sig:tmptest.fsi tmptest.fs' in directory 'D:\j\workspace\release_ci_pa---866fd2c3\tests\fsharp\core\members\basics'. ERRORLEVEL 1 ERRORLEVEL 1
at NUnitConf.checkTestResult(Result`2 result) in D:\j\workspace\release_ci_pa---866fd2c3\tests\fsharp\nunitConf.fs:line 14

2) Failed : FSharp-Tests-Core+Members+Basics.Basics(FSI_FILE)
Error running command 'D:\j\workspace\release_ci_pa---866fd2c3\tests\..\release\net40\bin\fsiAnyCPU.exe' with args ' -r:System.Core.dll --nowarn:20 --define:INTERACTIVE --maxerrors:1 --abortonerror test.fsi test.fs' in directory 'D:\j\workspace\release_ci_pa---866fd2c3\tests\fsharp\core\members\basics'. ERRORLEVEL 1 ERRORLEVEL 1
at NUnitConf.checkTestResult(Result`2 result) in D:\j\workspace\release_ci_pa---866fd2c3\tests\fsharp\nunitConf.fs:line 14

3) Failed : FSharp-Tests-Core+Members+Basics.Basics(FSC_OPT_PLUS_DEBUG)
Error running command 'D:\j\workspace\release_ci_pa---866fd2c3\tests\..\release\net40\bin\fsc.exe' with args ' -r:System.Core.dll --nowarn:20 --define:COMPILED --optimize+ --debug -o:test--optplus--debug.exe -g test.fsi test.fs' in directory 'D:\j\workspace\release_ci_pa---866fd2c3\tests\fsharp\core\members\basics'. ERRORLEVEL 1 ERRORLEVEL 1
at NUnitConf.checkTestResult(Result`2 result) in D:\j\workspace\release_ci_pa---866fd2c3\tests\fsharp\nunitConf.fs:line 14

4) Failed : FSharp-Tests-Core+Members+Basics.Basics(AS_DLL)
Error running command 'D:\j\workspace\release_ci_pa---866fd2c3\tests\..\release\net40\bin\fsc.exe' with args ' -r:System.Core.dll --nowarn:20 --define:COMPILED --optimize -a -o:test--optimize-lib.dll -g test.fsi test.fs' in directory 'D:\j\workspace\release_ci_pa---866fd2c3\tests\fsharp\core\members\basics'. ERRORLEVEL 1 ERRORLEVEL 1
at NUnitConf.checkTestResult(Result`2 result) in D:\j\workspace\release_ci_pa---866fd2c3\tests\fsharp\nunitConf.fs:line 14

5) Failed : FSharp-Tests-Core+Members+Basics.BasicsHw(FSI_FILE)
Error running command 'D:\j\workspace\release_ci_pa---866fd2c3\tests\..\release\net40\bin\fsiAnyCPU.exe' with args ' -r:System.Core.dll --nowarn:20 --define:INTERACTIVE --maxerrors:1 --abortonerror test.fsx' in directory 'D:\j\workspace\release_ci_pa---866fd2c3\tests\fsharp\core\members\basics-hw'. ERRORLEVEL 1 ERRORLEVEL 1
at NUnitConf.checkTestResult(Result`2 result) in D:\j\workspace\release_ci_pa---866fd2c3\tests\fsharp\nunitConf.fs:line 14

6) Failed : FSharp-Tests-Core+Members+Basics.BasicsHw(FSC_OPT_PLUS_DEBUG)
Error running command 'D:\j\workspace\release_ci_pa---866fd2c3\tests\..\release\net40\bin\fsc.exe' with args ' -r:System.Core.dll --nowarn:20 --define:COMPILED --optimize+ --debug -o:test--optplus--debug.exe -g test.fsx' in directory 'D:\j\workspace\release_ci_pa---866fd2c3\tests\fsharp\core\members\basics-hw'. ERRORLEVEL 1 ERRORLEVEL 1
at NUnitConf.checkTestResult(Result`2 result) in D:\j\workspace\release_ci_pa---866fd2c3\tests\fsharp\nunitConf.fs:line 14

7) Failed : FSharp-Tests-Core+Members+Basics.BasicsHwMutrec(FSI_FILE)
Error running command 'D:\j\workspace\release_ci_pa---866fd2c3\tests\..\release\net40\bin\fsiAnyCPU.exe' with args ' -r:System.Core.dll --nowarn:20 --define:INTERACTIVE --maxerrors:1 --abortonerror test.fs' in directory 'D:\j\workspace\release_ci_pa---866fd2c3\tests\fsharp\core\members\basics-hw-mutrec'. ERRORLEVEL 1 ERRORLEVEL 1
at NUnitConf.checkTestResult(Result`2 result) in D:\j\workspace\release_ci_pa---866fd2c3\tests\fsharp\nunitConf.fs:line 14

8) Failed : FSharp-Tests-Core+Members+Basics.BasicsHwMutrec(FSC_OPT_PLUS_DEBUG)
Error running command 'D:\j\workspace\release_ci_pa---866fd2c3\tests\..\release\net40\bin\fsc.exe' with args ' -r:System.Core.dll --nowarn:20 --define:COMPILED --optimize+ --debug -o:test--optplus--debug.exe -g test.fs' in directory 'D:\j\workspace\release_ci_pa---866fd2c3\tests\fsharp\core\members\basics-hw-mutrec'. ERRORLEVEL 1 ERRORLEVEL 1
at NUnitConf.checkTestResult(Result`2 result) in D:\j\workspace\release_ci_pa---866fd2c3\tests\fsharp\nunitConf.fs:line 14

9) Failed : FSharp-Tests-Typecheck+Sigs.sigs
Error running command 'D:\j\workspace\release_ci_pa---866fd2c3\tests\fsharpqa\testenv\bin\diff.exe' with args 'neg45.err neg45.bsl normalize' in directory 'D:\j\workspace\release_ci_pa---866fd2c3\tests\fsharp\typecheck\sigs'. ERRORLEVEL 1 ERRORLEVEL 1
at NUnitConf.checkTestResult(Result`2 result) in D:\j\workspace\release_ci_pa---866fd2c3\tests\fsharp\nunitConf.fs:line 14
```

**Comment by gmpl** [comment]
Thanks for your feedback @dsyme  I had a quick look at those tests but was not able to follow the code, actually I have no clue what is being tested (and failing) there, the error messages doesn't tell that much. Since I have no time in the week I will have a deeper look in the coming weekend.

**Comment by drvink** [comment]
@gmpl You need to look at the test output logs in order to determine what actually failed. For example, for the first test, `FSharp-Tests-Core+Members+Basics.Basics(GENERATED_SIGNATURE)`, the output is [here](http://dotnet-ci.cloudapp.net/job/Microsoft_visualfsharp/job/master/job/release_ci_part2_windows_nt_prtest/1254/artifact/tests/TestResults/FSharpNunit_Output.log/*view*/) (if you're wondering how I found that, it's [`FSharpNunit_Output.log`](http://dotnet-ci.cloudapp.net/job/Microsoft_visualfsharp/job/master/job/release_ci_part2_windows_nt_prtest/1254/) from the [`Windows_NT_Release_ci_part2_Build`](http://dotnet-ci.cloudapp.net/job/Microsoft_visualfsharp/job/master/job/release_ci_part2_windows_nt_prtest/1254/) CI results linked in this PR.

From that log, we can see that [the line that caused the test to fail](https://github.com/Microsoft/visualfsharp/blob/73f67ab53d3cd113e58429316653a46810e2e126/tests/fsharp/core/members/basics/test.fs#L1805) had this error:
```
tmptest.fs(1805,42): error FS0071: Type constraint mismatch when applying the default type 'DateTime' for a type inference variable. This expression was expected to have type
    'DateTime'    
but here has type
    'TimeSpan'     Consider adding further type constraints
```

Your changes definitely cause different results from the inference algorithm here, which we can confirm for ourselves:

```
(* subtracting two DateTimes results in a TimeSpan *)
> let ts = DateTime.Now - DateTime.Now;;
val ts : TimeSpan = 00:00:00

(* this won't work *)
> ((DateTime.Now - DateTime.Now) : DateTime);;

  ((DateTime.Now - DateTime.Now) : DateTime);;
  -----------------^^^^^^^^^^^^

stdin(6,18): error FS0001: This expression was expected to have type
    DateTime    
but here has type
    TimeSpan    

(* subtracting a TimeSpan from a DateTime produces a DateTime *)
> ((DateTime.Now - ts) : DateTime);;
val it : DateTime = 2016/09/14 3:20:01 {Date = 2016/09/14 0:00:00;
                                        Day = 14;
                                        DayOfWeek = Wednesday;
                                        DayOfYear = 258;
                                        Hour = 3;
                                        Kind = Local;
                                        Millisecond = 52;
                                        Minute = 20;
                                        Month = 9;
                                        Second = 1;
                                        Ticks = 636094200010529801L;
                                        TimeOfDay = 03:20:01.0529801;
                                        Year = 2016;}

(* the inference result without this PR for line 1805: *)
let f2 (x:DateTime) y : DateTime = x - y;;
val f2 : x:DateTime -> y:TimeSpan -> DateTime

(* and line 1806: *)
> let g2 (x:DateTime) y : TimeSpan = x - y;;
val g2 : x:DateTime -> y:DateTime -> TimeSpan
```

**Comment by gmpl** [comment]
@drvink  Thanks a lot for the explanation.
Before doing this pull request I simplified a lot my original code which was more invasive than this.
I'm afraid I didn't test it so well, my fault, I wish I had infinite time.
But the good news is that apparently my original code pass those tests, I think I oversimplified it and broke the type inference in the last step.
So don't lose the hope, I will need some more weekends to clean-up the working code or find what's wrong with this simplification ... but in the end it will work  ;)

File: src/fsharp/ConstraintSolver.fs [hunk 1]
```diff
@@ -2136,31 +2136,39 @@ and ResolveOverloading
...
+                | Some traitInfo -> RecordMemberConstraintSolution csenv.SolverState m (WithTrace newTrace) traitInfo (TTraitSolved (calledMeth.Method, calledMeth.CalledTyArgs))                
+                ++ (fun _ ->
                      CanMemberSigsMatchUpToCheck 
                          csenv 
                          permitOptArgs 
                          alwaysCheckReturn
                          (MustUnifyInsideUndo csenv ndeep newTrace) 
                          (TypesMustSubsumeOrConvertInsideUndo csenv ndeep (WithTrace newTrace) m)
-                         (ArgsEquivInsideUndo csenv Trace.New isConstraint) 
+                         (ArgsEquivInsideUndo csenv Trace.New cx.IsSome)
```
Comment (dsyme) [inline src/fsharp/ConstraintSolver.fs]: This use of ``Trace.New`` in the original code looks suspicious.  It feels like it should certainly be ``(WithTrace newTrace)``

Comment (gmpl) [inline src/fsharp/ConstraintSolver.fs]: Good point @dsyme, looks like a bug. Because if a new trace is used then <code>FilterEachThenUndo</code> will not undo those changes.
Now if this is a bug that has no consequences I wonder if that operation must be inside an undo at all.

Comment (gmpl) [inline src/fsharp/ConstraintSolver.fs]: Yes, actually the trace is not used at all (see the function <code>ArgsEquivInsideUndo</code>).

Comment (dsyme) [inline src/fsharp/ConstraintSolver.fs]: OK, thanks, pease remove the argument as part of this PR

Comment (gmpl) [inline src/fsharp/ConstraintSolver.fs]: No worries, I will.

**Comment by gmpl** [comment]
Now the adjustments are not that small if you look at the number of lines changed but in fact the logic change is still small, I just introduced an additional argument to pass the member constraint solution and apply it later, when all the inference equations are already in place.

**Comment by forki** [comment]
Didn't you have a extra repro sample? Can you please add it as a test?

File: src/fsharp/ConstraintSolver.fs [hunk 2]
```diff
@@ -737,6 +737,10 @@ and SolveTypEqualsTyp (csenv:ConstraintSolverEnv) ndeep m2 (trace: OptionalTrace
     let sty1 = stripTyEqnsA csenv.g canShortcut ty1
     let sty2 = stripTyEqnsA csenv.g canShortcut ty2
 
+    match cxsln with
+    | Some (traitInfo, traitSln) when traitInfo.Solution.IsNone -> TransactMemberConstraintSolution traitInfo trace traitSln
```
Comment (forki) [inline src/fsharp/ConstraintSolver.fs]: Is this a imperative call?

Comment (gmpl) [inline src/fsharp/ConstraintSolver.fs]: Yes, it is writing to a reference cell, see [this comments](https://github.com/gmpl/visualfsharp/blob/aec3cea9001aff7947770e88c31db48913e05714/src/fsharp/ConstraintSolver.fs#L1286) from the original code.

**Comment by gmpl** [comment]
@forki Do you mean as a test constrained by the compile time? If so, can you point me to similar tests already in the solution?

**Comment by forki** [comment]
@gmpl I think I meant the sample given in https://github.com/Microsoft/visualfsharp/issues/351 - does your pr fix it as well? If so then please try to add it as a test.

**Comment by gmpl** [comment]
@forki It doesn't "fix" the results. They are the same but it compiles it 20 times faster. I'm not sure if there are tests over the compile time in the solution.

**Comment by gmpl** [comment]
@dsyme Another alternative is rather than adding an additional parameter to the function, add a property in the Solver State to store the constraint solution until we commit it. That change will be quite small in terms of lines of code and function signatures. What do you think?

**Comment by forki** [comment]
@gmpl I will try to run that experiment: https://github.com/Microsoft/visualfsharp/pull/1550

**Comment by forki** [comment]
@gmpl can you please merge #1550 into this PR? Then I can close it.

**Comment by gmpl** [comment]
@forki Now I see you were referring to another issue, I was thinking in the other one but they were marked as duplicate. So yes I will have a look tonight and try to merge it. Please leave it open for today, thanks.

**Comment by gmpl** [comment]
@forki Do you know why all test failed after merging your test?
I mean, I can see the error message, but I wonder why it was passing before, you just added code for that specific test.

**Comment by forki** [comment]
Unable to find version '14.0.24516-Pre' of package 'VisualCppTools'.

that's nuget hickup, completely unrelatd to the merge.

@dotnet-bot test this please.

**Comment by forki** [comment]
@maartenba sorry for asking you again. ;-)

00:36:29   NotFound https://api.nuget.org/v3-flatcontainer/visualcpptools/14.0.24516-pre/visualcpptools.14.0.24516-pre.nupkg 257ms
00:36:29   NotFound https://myget-2e16.kxcdn.com/artifacts/fsharp-daily/nuget/v3/flatcontainer/visualcpptools/14.0.24516-pre/visualcpptools.14.0.24516-pre.nupkg 683ms

WARNING: Unable to find version '14.0.24516-Pre' of package 'VisualCppTools'.


Any ideas why this thing is not available?

**Comment by forki** [comment]
ok just saw @KevinRansom was already fixing this in master. https://github.com/Microsoft/visualfsharp/commit/4d87596bd24f956b4599ed15f1c4cfcf321086de

so @gmpl please rebase this PR

**Comment by forki** [comment]
mmm I tried to rebase https://github.com/Microsoft/visualfsharp/pull/1552 - but get the same error. so seems master is broken as well.

**Comment by forki** [comment]
ok. last hope is @maartenba

**Comment by maartenba** [comment]
@forki which feed did this package come from?

**Comment by forki** [comment]
I have no idea. I tried to find out by looking into older builds but without luck

**Comment by forki** [comment]
https://blogs.msdn.microsoft.com/vcblog/2016/02/16/try-out-the-latest-c-compiler-toolset-without-waiting-for-the-next-update-of-visual-studio/ suggest it was available on nuget.org

**Comment by forki** [comment]
sorry. it says:

**Comment by forki** [comment]
but that's gone as well!?

**Comment by dsyme** [comment]
@gmpl Could you explain more why it's correct to put [this logic](https://github.com/Microsoft/visualfsharp/pull/1530/files#diff-641da7c52ffc888699888ebd03559cb5R740) in SolveTypeEqualsType?  Why there? 

For example, SolveTypeEqualsType has a shortcut case  ``if ty1 === ty2 then CompleteD else`` based on pointer equality of two types.  But the code you've added won't be hit on that path, will it?  

I just don't yet grok why SolveTypeEqualsType is the right place, thanks

**Comment by gmpl** [comment]
@dsyme Logically the solution would be committed at the beginning of the selection of the method, that's what I did on the first commit but it didn't work for all scenarios.
Why it didn't work?
I don't understand all the details, I wish there was more documentation about the constraint solver, but I know that in some situations resolving a type early will mislead to unsolved types, typical examples are when there are few overloads of a method.
So the solver needs to pretend it doesn't know the solution for the constraint at the beginning, in order to come up with more type equations, then the question is where do we commit the solution? Well, it has to be somewhere before calling to <code>ResolveOverloading</code> again, ideally before solving and adding more constraints, that place is <code>SolveTyparEqualsTyp</code> right before <code>SolveRelevantMemberConstraintsForTypar</code> is called.
But since that function is only called by <code>SolveTypEqualsTyp</code> I think it's safe to commit the solution there.
I tried adding it at the very beginning of the function body, but I have a (very exotic) [test code](https://github.com/gmpl/FsControl/blob/master/FsControl.Core/Samples/Haskell.fsx#L560) that failed, moving it right before the shortcut case you mention solved the issue without slowing back down the other scenarios. So it can be (and maybe should be) 3 lines more up in the code but not before the shortcut. I guess this is because when the types are the same it returns to the caller without having solved anything else but with the constraint solution committed, which as I tried explaining is misleading.

**Comment by forki** [comment]
@gmpl please rebase on master and force push in order to solve the build issue

**Comment by gmpl** [comment]
@dsyme Here's another viewpoint, maybe that's not the best place and it should be as you suggested, before the shortcut.

The fact that just a piece of very exotic code was compiling before without type annotations shouldn't prevent us to write those lines in the right place.

Personally I'm fine if that code breaks from now on, anyway it also breaks with a minimal change in the definition of <code><*></code> , so it might be working just by chance. And I don't think you will find code like that in production projects.

For that particular case we can work later on another PR, because there are many cases like that not compiling and that's something that can be improved in the type inference.

**Comment by dsyme** [comment]
@gimpl What was the example code that broke when your insertion was before the shortcut line? Thanks

Re needing to do things late - yes, I imagine that in method overload resolution there are some pathological cases where operations are not fully monotonic w.r.t. adding constraint solutions - so for example adding a new constraint solution (which was eventually being deduced originally) earlier leads to the solver going down paths it wouldn't have otherwise gone down, and deriving conflicting information.

**Comment by gmpl** [comment]
@dsyme Yes, that's probably the case of the example code.
I linked it in my comment, it's somewhere in a script in FsControl, in [this line](https://github.com/gmpl/FsControl/blob/master/FsControl.Core/Samples/Haskell.fsx#L560).
Also [this line](https://github.com/gmpl/FsControl/blob/master/FsControl.Core/Samples/Haskell.fsx#L577) fails and if you play a bit with those idioms and generic code most likely you will find more cases, but as I said I'm fine if that code breaks from now on.
I will try to come up with a minimal repro for those cases.
Apart from those pathological cases I didn't find any other piece of code that breaks, and I tested it with many different and exotic snippets.

**Comment by gmpl** [comment]
@forki  I rebased last night and Windows_NT Release_ci_part1 Build failed. Just did it again and now is AppVeyor the problem.
None of these problems seems to be related to this enhancement.

**Comment by gmpl** [comment]
@dsyme Here's the [minimal repro](https://gist.github.com/gmpl/453e55a0a615644561fbec1a0a771f62).

Now this is even more exotic than what I thought. I tested the FsControl sample file I sent you before with the "before shortcut version" and it works.

I investigated what was the difference with my local script and found that my local script has a <code><*></code> definition which doesn't use overloads.

So, if you use the "full version" works fine.

I also did some other tests and found that the "before shortcut version" speeds up even more (but just a little bit) some local tests I have.

Now this convinced me that we should commit the solution before the shortcut and investigate this corner cases as a separate thing.

**Comment by drvink** [comment]
Huge thanks to @gmpl and @dsyme for this. I'm so happy this is fixed!

**Comment by forki** [comment]
excellent news!

**Comment by cloudRoutine** [comment]
@gmpl since you're the new typechecking pro maybe you can put this [parallel typechecking algorithm](http://homes.soic.indiana.edu/samth/parallel-typecheck-draft.pdf) to work for us?

**Comment by vasily-kirichenko** [comment]
I'm trying compiler built from master and there are some errors on my large solution. The first one:
```fsharp
// in Processes.fs
namespace Foo

type ProcessError =
    ...
    | CannotStart of string * exn
    | DoesNotExist of PID

// in Supervisor.fs, which lays below Processes.fs
namespace Foo

type DeadStatus =
    | DoesNotExist
    | CannotLaunch
    ...
```
(note that both types have `DoesNotExist case, but one of them carries `PID` value).

It compiles OK with F# 4, but does not with master:

(`open Foo ...`)
image

I thing name resolution order has changed - old compiler picked _first_ definition among symbols with same name while the new one picks _last_ one.

**Comment by vasily-kirichenko** [comment]
After fixing compilation errors, the timings are as following:
```

**Comment by forki** [comment]
wow

**Comment by vasily-kirichenko** [comment]
@forki half of projects are C#, so the speedup on a pure F# solutions will be even more significant.

**Comment by forki** [comment]
that's amazing

**Comment by 7sharp9** [comment]
I wonder if it will have much difference on type checking during IDE, I guess it needs to be back ported to FCS to test.

**Comment by vasily-kirichenko** [comment]
AFAIK this PR speed ups the type checker specifically, so FCS should speed up even further as it does not do byte code generation.

**Comment by cloudRoutine** [comment]
hopefully parallelism can speed it up to the point where people stop complaining about how slow the compiler is 😸

**Comment by vasily-kirichenko** [comment]
@cloudRoutine for IDE? Yes, I hope. As for building, it's already parallized by msbuild/VS, so CPU is about 100% load.

**Comment by cloudRoutine** [comment]
@vasily-kirichenko  no I mean **[parallelizing the actual process of typechecking](http://homes.soic.indiana.edu/samth/parallel-typecheck-draft.pdf)**

**Comment by gmpl** [comment]
@cloudRoutine That would be very interesting, unfortunately I can only work on weekends.
Anyway I'm still not happy with the code, I think before going parallel we should make the solver algorithm more consistent.
I'm convinced it can speed up some more and it can also be smarter, there are many cases where it's not able to solve and should be possible.

**Comment by gmpl** [comment]
@vasily-kirichenko I don't think this optimization changed the name resolution order, there were other important changes since F# 4.0 that most likely did that. Did you check with the previous version before this commit?

Nice to hear that your project compilation speeds up a lot, do you use FsControl or a do you use a lot of this kind of overloaded functions there?

I tried to compile [FsControl](https://github.com/gmpl/FsControl), before it was taking 2' 30'' and now with this version it takes only 30 secs :) Anyway I think the FsControl.dll compilation is not the part that will benefit the most from this optimization.
The sample scripts should speed up more, I still have to test and compare them.

**Comment by dsyme** [comment]
@gmpl @forki @vasily-kirichenko It's crucial that we determine where name resolution changed in F# 4.1 as soon as possible, and if possible revert it to previous behaviour.

@gmpl @vasily-kirichenko If there's any chance you could do a binary chop on the commits for F# 4.1 to determine where name resolution changed that would be awesome.  It might be that using the commit history for http://github.com/fsharp/fsharp may be easier.

@forki Could you carefully check if your name resolution changes altered the order of name resolution results in cases like the above

**Comment by forki** [comment]
@vasily-kirichenko can you come up with a repro case?

**Comment by vasily-kirichenko** [comment]
https://github.com/Microsoft/visualfsharp/pull/1530#issuecomment-254467693?

**Comment by forki** [comment]
@vasily-kirichenko I tried to put that into a test: #1652

Looks like it resolves fine for me on master:

image

**Comment by vasily-kirichenko** [comment]
Maybe it behaves differently if files compiled as part of a project?

**Comment by vasily-kirichenko** [comment]
Does it work the same in F# 4 FSI?

**Comment by dsyme** [comment]
@forki It might need to have the two namespace fragments in one file

Come to think of it, I think this may have been a deliberate bug fix, see [this commit](https://github.com/Microsoft/visualfsharp/commit/bdb842797677db4893fe54957476caaef15b2b01) - I do recall fixing an inconsistency when combining namespace fragments from a single file.  It might even have been that the effective combined declaration order flipped every time a "combine" happened.

If so I suppose we would be OK with this being a breaking change

**Comment by forki** [comment]
Tbh I think this case was always the same and I think the current behavior is correct. @vasily-kirichenko can you try to modify my sample so that it breaks for you?

**Comment by vasily-kirichenko** [comment]
@forki I tried to reproduce it on a minimal project and failed. Also I benchmarked 4.0/4.1-before-optimization/4.1-master compilers by compiling FCS project and all of them showed identical results (~42 seconds). Now I'm not sure why my large solution is compiled more than 2x faster on master compiler. I switched between 4.0 and master ones several times and it's reliably reproduced. Can it be some mess with ngen or the like?

**Comment by gmpl** [comment]
@vasily-kirichenko if your solution doesn't make use of constrained overload resolution I don't see why it would speed up that much with this PR. But if it does, even in a relative small number it could have improved, since some cases were taking insane compile time.
